sphinx/pycode/parser.py

```diff
@@ -295,12 +295,9 @@ def add_variable_annotation(self, name: str, annotation: ast.AST) -> None:
             self.annotations[basename, name] = ast_unparse(annotation)
 
     def is_final(self, decorators: list[ast.expr]) -> bool:
-        final = []
-        if self.typing:
-            final.append('%s.final' % self.typing)
-        if self.typing_final:
-            final.append(self.typing_final)
-
+        final = self.typing_final_names.copy()
```

Comment:
Can you explain the `.copy()`? Previously it always created an empty list.

Reply:
`typing_final` was a `str | None` before, I changed it to a list to store multiple seen names. This copies to avoid mutation. This is just a simplification compared to making an empty list, then extending it immediately.

Reply:
I'd prefer to be explicit and use an empty list. The `.copy()` makes local reasoning harder because the instance state might have been mutated elsewhere.

Reply:
Not sure what you mean by mutation elsewhere, unless other code made it not a list. Could also do `list(self.typing_final_names)`. Using an empty list would be something like:
```py
final = []
for name in self.typing_final_names:
    final.append(name)
```
Which is pointlessly overcomplicated.

Reply:
> Not sure what you mean by mutation elsewhere, unless other code made it not a list

Mainly the risk of having more entries than needed, becuase this is a list rather than a set. However I think making it a set would work?:

```
final = set(self.typing_final_names) | {f'{mod}.final' for mod in self.typing_mods}
```

Reply:
Ah, yes it can be possible for the list to have multiple copies, but only if the user did imports multiple times. Changing everything to sets would fix that, I just didn't think that was common enough to be worth using sets, when they'd normally only hold up to 4 items.